src/python/python-proto/python_proto/serde.py

```diff
@@ -0,0 +1,83 @@
+from __future__ import annotations
+
+from abc import ABCMeta, abstractmethod, abstractstaticmethod
+from typing import Generic, TypeVar
+
+from google.protobuf.message import Message as _Message
+
+P = TypeVar("P", bound=_Message)
+
+
+class SerDe(Generic[P], metaclass=ABCMeta):
+    proto_cls: type[P]
+
+    @classmethod
+    def __subclasshook__(cls, subclass: SerDe[P]) -> bool:
+        return (
+            hasattr(subclass, "proto_cls")
+            and hasattr(subclass, "deserialize")
+            and callable(subclass.deserialize)
+            and hasattr(subclass, "serialize")
+            and callable(subclass.serialize)
+            and hasattr(subclass, "from_proto")
+            and callable(subclass.from_proto)
+            and hasattr(subclass, "into_proto")
+            and callable(subclass.into_proto)
+        )
```

Comment:
We should probalby just use pydantic if we want to do this sort of thing, but this is fine for now

Reply:
this is all pre-existing and i dont want to change it :')